### PR TITLE
Fix nonsense non-tupled `Fn` trait error

### PR DESCRIPTION
**Pull request by compiler-errors**
Given this code:

```rust
#![feature(unboxed_closures)]

fn a<F: Fn<usize>>(f: F) {}

fn main() {
    a(|_: usize| {});
}
```

We currently emit this error:
```
error[E0631]: type mismatch in closure arguments
 --> src/main.rs:6:5
  |
6 |     a(|_: usize| {});
  |     ^ ---------- found signature of `fn(usize) -> _`
  |     |
  |     expected signature of `fn(usize) -> _`
  |
note: required by a bound in `a`
 --> src/main.rs:3:9
  |
3 | fn a<F: Fn<usize>>(f: F) {}
  |         ^^^^^^^^^ required by this bound in `a`

For more information about this error, try `rustc --explain E0631`.
error: could not compile `playground` due to previous error
```
Notably, it says the same thing for "expected" and "found"!

Fix the output so that we instead emit:
```
error[E0308]: mismatched types
 --> /home/gh-compiler-errors/test.rs:6:5
  |
6 |     a(|_: usize| {});
  |     ^ types differ
  |
  = note: expected trait `Fn<usize>`
             found trait `Fn<(usize,)>`
note: required by a bound in `a`
 --> /home/gh-compiler-errors/test.rs:3:9
  |
3 | fn a<F: Fn<usize>>(f: F) {}
  |         ^^^^^^^^^ required by this bound in `a`

error: aborting due to previous error
```

The error could still use some work, namely the "mismatched types" part, but I'm leaving it a bit rough since the only way you'd ever get this error is when you're messing with `#![feature(unboxed_closures)]`. 

Simply making sure we actually print out the difference in trait-refs is good enough for me. I could probably factor in some additional improvements if those are desired.

### CODE REVIEW

**Review by cjgillot — COMMENTED** [review]
Sorry for the delay @compiler-errors.
When we have `Fn<A>` with a concrete type `A`, we should probably report that a tuple is expected. The diagnostic improvement for generic `A` is great though.

File: src/test/ui/unboxed-closures/non-tupled-arg-mismatch.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#![feature(unboxed_closures)]
+
+fn a<F: Fn<usize>>(f: F) {}
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply src/test/ui/unboxed-closures/non-tupled-arg-mismatch.rs:3]: Well making this a hard error #99943, so I don't want to just introduce a special case logic here.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -940,20 +941,43 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
                 self.reported_closure_mismatch.borrow_mut().insert((span, found_span));
 
+                let mut not_tupled = false;
+
                 let found = match found_trait_ref.skip_binder().substs.type_at(1).kind() {
                     ty::Tuple(ref tys) => vec![ArgKind::empty(); tys.len()],
-                    _ => vec![ArgKind::empty()],
+                    _ => {
+                        not_tupled = true;
+                        vec![ArgKind::empty()]
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:950]: No particular reason, I can fix that I guess.

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:950]: Oh, I know why. This code also handles `Generator<T>` which takes a generic argument that _isn't_ a tupled set of arguments.

File: src/test/ui/unboxed-closures/non-tupled-arg-mismatch.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (compiler-errors) [reply src/test/ui/unboxed-closures/non-tupled-arg-mismatch.rs:3]: Also, reporting that a tuple is expected and bailing out is not gonna work when you have a generic `Fn<T>` where `T` is a type param.

**Comment by compiler-errors** [comment]
I actually don't know if I have anything to change here. See replied comments.

@rustbot ready

**Comment by matthiaskrgr** [comment]
@bors r-
````
     Checking rustc_session v0.0.0 (/checkout/compiler/rustc_session)
    Checking rustc_attr v0.0.0 (/checkout/compiler/rustc_attr)
    Checking rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
    Checking rustc_middle v0.0.0 (/checkout/compiler/rustc_middle)
    Checking rustc_ast_passes v0.0.0 (/checkout/compiler/rustc_ast_passes)
    Checking rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error: passing `TypeError<'_>` by reference
  --> compiler/rustc_middle/src/ty/error.rs:78:29
   |
78 |     pub fn involves_regions(&self) -> bool {
   |                             ^^^^^ help: try passing by value: `TypeError<'_>`
   |
   = note: `-D rustc::pass-by-value` implied by `-D warnings`

error: could not compile `rustc_middle` due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `rustc_middle` due to previous error
Build completed unsuccessfully in 0:02:05
````

https://github.com/rust-lang/rust/pull/100593#issuecomment-1215533202

**Pull request merged**